spec/models/message_summarizer_spec.rb

```diff
@@ -0,0 +1,226 @@
+
+      it { expect(subject.order_details_in_dispute.count).to eq(1) }
+
+      it "returns only disputed order detail messages when iterating" do
+        subject.each do |message_summary|
+          expect(message_summary)
+            .to be_kind_of(MessageSummarizer::OrderDetailsInDispute)
+        end
+      end
+
+      context "when not scoped to a facility" do
+        let(:current_facility) { nil }
+
+        it_behaves_like "there are no messages"
+      end
+    end
+
+    context "and the user cannot access disputed order details" do
+      it_behaves_like "there are no messages"
+    end
+  end
+
+  context "when a problem order detail exists" do
+    let(:product) { create(:setup_item) }
+
+    before(:each) { set_problem_order }
+
+    context "and the user can access problem orders" do
+      before { ability.can(:show_problems, Order) }
+
+      it_behaves_like "there is one overall message"
+
+      it { expect(subject.problem_order_details.count).to eq(1) }
+
+      it "returns only problem order detail messages when iterating" do
+        subject.each do |message_summary|
+          expect(message_summary)
+            .to be_kind_of(MessageSummarizer::ProblemOrderDetails)
+        end
+      end
+
+      context "when not scoped to a facility" do
+        let(:current_facility) { nil }
+
+        it_behaves_like "there are no messages"
+      end
+    end
+
+    context "and the user cannot access problem orders" do
+      it_behaves_like "there are no messages"
+    end
+  end
+
+  context "when a problem reservation order detail exists" do
+    before(:each) { set_problem_order }
+
+    context "and the user can access problem reservations" do
+      before { ability.can(:show_problems, Reservation) }
+
+      it_behaves_like "there is one overall message"
+
+      it { expect(subject.problem_reservation_order_details.count).to eq(1) }
+
+      it "returns only problem reservation messages when iterating" do
+        subject.each do |message_summary|
+          expect(message_summary)
+            .to be_kind_of(MessageSummarizer::ProblemReservationOrderDetails)
+        end
+      end
+
+      context "when not scoped to a facility" do
+        let(:current_facility) { nil }
+
+        it_behaves_like "there are no messages"
+      end
+    end
+
+    context "and the user cannot access problem reservations" do
+      it_behaves_like "there are no messages"
+    end
+  end
+
+  context "when a training request exists" do
+    before { create(:training_request, product: product) }
+
+    context "and the user can manage training requests" do
+      before { ability.can(:manage, TrainingRequest) }
+
+      it_behaves_like "there is one overall message"
+
+      it { expect(subject.training_requests.count).to eq(1) }
+
+      it "returns only training requests when iterating" do
+        subject.each do |message_summary|
+          expect(message_summary)
+            .to be_kind_of(MessageSummarizer::TrainingRequests)
```

Comment:
I think I'd like to see some deeper testing, like seeing that the message ends up as "Training Requests (1)".